File: Shared/Models/DelayStatus.swift
import SwiftUI

// Estado de demora de un arribo. minor vs major se decide con la tolerancia del ramal.
enum DelayStatus: Equatable, Hashable {
    case onTime
    case early(seconds: Int)
    case minor(seconds: Int)
    case major(seconds: Int)
    case noData
    case cancelled

    private static func minutes(_ seconds: Int) -> Int {
        Int((Double(abs(seconds)) / 60).rounded())
    }

    var label: String {
        switch self {
        case .onTime:            return "En horario"
        case .early(let s):      return "-\(Self.minutes(s)) min"
        case .minor(let s):      return "+\(Self.minutes(s)) min"
        case .major(let s):      return "+\(Self.minutes(s)) min"
        case .noData:            return "Sin datos"
        case .cancelled:         return "Cancelado"
        }
    }

    var shortLabel: String {
        switch self {
        case .onTime:            return "En hora"
        case .early(let s):      return "-\(Self.minutes(s))"
        case .minor(let s):      return "+\(Self.minutes(s))"
        case .major(let s):      return "+\(Self.minutes(s))"
        case .noData:            return "S/D"
        case .cancelled:         return "Canc."
        }
    }

    var color: Color {
        switch self {
        case .onTime, .early:    return Palette.onTime
        case .minor:             return Palette.minorDelay
        case .major, .cancelled: return Palette.majorDelay
        case .noData:            return Palette.noData
        }
    }

    // Segundos de demora con signo (negativo = adelantado). nil si no hay dato.
    var delaySeconds: Int? {
        switch self {
        case .early(let s):  return -s
        case .minor(let s):  return s
        case .major(let s):  return s
        case .onTime:        return 0
        case .noData, .cancelled: return nil
        }
    }
}

File: android/app/src/main/java/com/alandaitch/anden/data/model/DelayStatus.kt
package com.alandaitch.anden.data.model

import androidx.compose.ui.graphics.Color
import com.alandaitch.anden.ui.theme.Palette
import kotlin.math.abs
import kotlin.math.roundToInt

// Estado de demora de un arribo. minor vs major se decide con la tolerancia del ramal.
sealed class DelayStatus {
    object OnTime : DelayStatus()
    data class Early(val seconds: Int) : DelayStatus()
    data class Minor(val seconds: Int) : DelayStatus()
    data class Major(val seconds: Int) : DelayStatus()
    object NoData : DelayStatus()
    object Cancelled : DelayStatus()

    val label: String
        get() = when (this) {
            is OnTime -> "En horario"
            is Early -> "-${minutes(seconds)} min"
            is Minor -> "+${minutes(seconds)} min"
            is Major -> "+${minutes(seconds)} min"
            is NoData -> "Sin datos"
            is Cancelled -> "Cancelado"
        }

    val shortLabel: String
        get() = when (this) {
            is OnTime -> "En hora"
            is Early -> "-${minutes(seconds)}"
            is Minor -> "+${minutes(seconds)}"
            is Major -> "+${minutes(seconds)}"
            is NoData -> "S/D"
            is Cancelled -> "Canc."
        }

    val color: Color
        get() = when (this) {
            is OnTime, is Early -> Palette.onTime
            is Minor -> Palette.minorDelay
            is Major, is Cancelled -> Palette.majorDelay
            is NoData -> Palette.noData
        }

    // Segundos de demora con signo (negativo = adelantado). null si no hay dato.
    val delaySeconds: Int?
        get() = when (this) {
            is Early -> -seconds
            is Minor -> seconds
            is Major -> seconds
            is OnTime -> 0
            is NoData, is Cancelled -> null
        }

    companion object {
        private fun minutes(seconds: Int): Int = (abs(seconds).toDouble() / 60.0).roundToInt()
    }
}
